### PR TITLE
add script to preprocess data.

### PR DESCRIPTION
**Pull request by zheng-da**
This is an initial implementation of the data loading pipeline in a single machine. It provides the following support:
* text tokenization for the node/edge features.
* multi-processing to load nodes and edges in parallel.
* supports node/edge data stored in Parquet.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/graph_loading/construct_graph.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+    if '*' in in_files:
+        in_files = glob.glob(in_files)
+    elif not isinstance(in_files, list):
+        in_files = [in_files]
+    in_files.sort()
+    return in_files
+
+def parse_node_data(i, in_file, feat_ops, node_id_col, label_conf,
+                    read_file, return_dict):
+    """ Parse node data.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:314]: I'm not sure if this is user-facing function. but I have added comments to make the code more readable.

File: tools/graph_loading/partition_graph.py
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/partition_graph.py:1]: good point. I'll move it back.

File: tools/graph_loading/partition_graph.py [hunk 2]
```diff
@@ -32,7 +32,8 @@
     # dataset and file arguments
     argparser.add_argument("-d", "--dataset", type=str, required=True,
                            help="dataset to use")
-    argparser.add_argument("--filepath", type=str, default=None)
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/partition_graph.py:35]: let me reverse the change. but I think this argument is very confusing.

File: tools/graph_loading/construct_graph.py [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+import numpy as np
+from functools import partial
+import argparse
+
+from transformers import BertTokenizer
+import torch as th
+
+##################### The I/O functions ####################
+
+def read_data_parquet(data_file):
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:19]: Your question is why we need to create a function to read parquet file?
First, this is a simple wrapper function to read parquet file. Because parquet format doesn't support multi-dimension arrays, it also does some data transformation.

[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:19]: We'll need to add more functions to support other formats, such as JSON and CSV.

File: tools/graph_loading/construct_graph.py [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+        if d.dtype.hasobject:
+            d = [d[i] for i in range(len(d))]
+            d = np.stack(d)
+        data[key] = d
+    return data
+
+def write_data_parquet(data, data_file):
+    """ Write data in parquet files.
+
+    Normally, Parquet cannot support multi-dimension arrays.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:51]: I don't understand your comment or which part is confusing.
Parquet format doesn't have native support multi-dimension arrays. It can only support vectors. It has nothing to do with the data type of each individual element. I'm not sure if this is what you mean in the comment.

File: tools/graph_loading/construct_graph.py
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:1]: this is a script, instead of a library. do we need to worry about exposed API?

File: tools/graph_loading/construct_graph.py [hunk 5]
```diff
@@ -0,0 +1,680 @@
...
+
+    The function parses an edge file that contains the source and destination node
+    IDs, edge features and potentially edge labels. The edge file is parsed
+    according to users' configuration and performs some feature transformation
+    and save the result in `return_dict`.
+
+    Parameters
+    ----------
+    file_idx : int
+        The index of the edge file among all node files.
```
Comment (classicsong) [inline tools/graph_loading/construct_graph.py:358]:
```suggestion
        The index of the edge file among all edge files.
```

File: tools/graph_loading/construct_graph.py [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+        Whether or not to remap node IDs
+
+    Returns
+    -------
+    dict: node ID map
+    dict: node features.
+    """
+    node_data = {}
+    node_id_map = {}
+    for process_conf in process_confs:
```
Comment (classicsong) [inline tools/graph_loading/construct_graph.py:462]: Each process_conf corresponds to one node type, right?

File: tools/graph_loading/construct_graph.py [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+        type_node_id_map = [None] * len(return_dict)
+        type_node_data = {}
+        for i, (node_ids, data) in return_dict.items():
+            for feat_name in data:
+                if feat_name not in type_node_data:
+                    type_node_data[feat_name] = [None] * len(return_dict)
+                type_node_data[feat_name][i] = data[feat_name]
+            type_node_id_map[i] = node_ids
+
+        assert type_node_id_map[0] is not None
```
Comment (classicsong) [inline tools/graph_loading/construct_graph.py:498]: Why not assert type_node_id_map[i] is not None for i in len(type_node_id_map) ?

File: tools/graph_loading/construct_graph.py [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+        type_node_id_map = np.concatenate(type_node_id_map)
+        # We don't need to create ID map if the node IDs are integers,
+        # all node Ids are in sequence start from 0 and
+        # the user doesn't force to remap node IDs.
+        if np.issubdtype(type_node_id_map.dtype, np.integer) \
```
Comment (classicsong) [inline tools/graph_loading/construct_graph.py:503]: Are you assuming the node_id_maps are ordered in type_node_id_map? Is manager.dict() an sorted dict (key-val pairs are stored according to the order of keys) or not?

File: tools/graph_loading/construct_graph.py [hunk 9]
```diff
@@ -0,0 +1,680 @@
...
+                and np.all(type_node_id_map == np.arange(len(type_node_id_map))) \
+                and not remap_id:
+            num_nodes = len(type_node_id_map)
+            type_node_id_map = None
+        else:
+            num_nodes = len(type_node_id_map)
+            type_node_id_map = create_id_map(type_node_id_map)
+
+        for feat_name in type_node_data:
+            type_node_data[feat_name] = np.concatenate(type_node_data[feat_name])
```
Comment (classicsong) [inline tools/graph_loading/construct_graph.py:513]: Same here.

File: tools/graph_loading/construct_graph.py [hunk 10]
```diff
@@ -0,0 +1,680 @@
...
+        elif ntype in node_data:
+            for feat_name in node_data[ntype]:
+                # Here we only need to look at the length of the first node features
+                # to get the number of nodes for the node type.
+                num_nodes[ntype] = len(node_data[ntype][feat_name])
+                break
+        else:
+            # A node type must have either ID map or node data.
+            raise ValueError('Node type {} must have either ID map or node data'.format(ntype))
+    g = dgl.heterograph(edges, num_nodes_dict=num_nodes)
```
Comment (classicsong) [inline tools/graph_loading/construct_graph.py:662]: Are we going to add reverse edges?

File: tests/end2end-tests/data_process/data_gen.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
+import copy
+import json
+import os
+import pyarrow.parquet as pq
+import pyarrow as pa
+import numpy as np
+
+def write_data_parquet(data, data_file):
+    df = {}
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tests/end2end-tests/data_process/data_gen.py:9]: do we need docstring for test code?

File: tools/graph_loading/README.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+	]
+}
+```
+
+GraphStorm contains a script `construct_graph.py` that constructs a graph
+from the user's input data with the format described above. The script will save
+the constructed graph in a file with the DGL or DistDGL graph format. If users' data have
+string node IDs or the node IDs in the node files are not stored in an ascending order
+starting from 0, `construct_graph.py` will remap the node IDs and save the ID map
+into files. The node ID map of each node type is saved in separate files.
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/README.md:90]: do you mean the ID map file name or the content of the ID map file?

File: tools/graph_loading/construct_graph.py [hunk 13]
```diff
@@ -0,0 +1,680 @@
...
+    table = pq.read_table(data_file)
+    pd = table.to_pandas()
+    data = {}
+    for key in pd:
+        d = np.array(pd[key])
+        # For multi-dimension arrays, we split them by rows and
+        # save them as objects in parquet. We need to merge them
+        # together and store them in a tensor.
+        if d.dtype.hasobject:
+            d = [d[i] for i in range(len(d))]
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:43]: why do you think it's 2D array?

File: tools/graph_loading/construct_graph.py [hunk 14]
```diff
@@ -0,0 +1,680 @@
...
+    -------
+    dict : the key is the data name, the value is the processed data.
+    """
+    new_data = {}
+    for feat_col, feat_name, dtype, op in ops:
+        # If the transformation is defined on the feature.
+        if op is not None:
+            res = op(data[feat_col])
+            if isinstance(res, dict):
+                for key, val in res.items():
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:215]: i think we need to settle how to deal with this case. some preprocessing functions output multiple arrays instead of a single array. how do we assign names to these arrays.

File: tools/graph_loading/construct_graph.py [hunk 15]
```diff
@@ -0,0 +1,680 @@
...
+    assert 'task_type' in label_conf, "'task_type' must be defined in the label field."
+    if label_conf['task_type'] == 'classification':
+        label = np.int32(label)
+    if 'split_type' in label_conf:
+        train_split, val_split, test_split = label_conf['split_type']
+        rand_idx = np.random.permutation(len(label))
+        train_idx = rand_idx[0:int(len(label) * train_split)]
+        val_idx = rand_idx[int(len(label) * train_split):int(len(label) * (train_split + val_split))]
+        test_idx = rand_idx[int(len(label) * (train_split + val_split)):]
+        train_mask = np.zeros((len(label),), dtype=np.int8)
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:251]: we need. let's do it in another PR. We need to discuss how to define it.

File: tools/graph_loading/construct_graph.py [hunk 16]
```diff
@@ -0,0 +1,680 @@
...
+
+    Returns
+    -------
+    a list of str : the full name of input files.
+    """
+    if '*' in in_files:
+        in_files = glob.glob(in_files)
+    elif not isinstance(in_files, list):
+        in_files = [in_files]
+    in_files.sort()
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:309]: how to check a file is valid?

File: tools/graph_loading/construct_graph.py (hunk 6 above)
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:462]: yes

File: tools/graph_loading/construct_graph.py (hunk 8 above)
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:503]: i don't understand what you mean. i don't think we need to assume it's sorted.

File: tools/graph_loading/construct_graph.py [hunk 17]
```diff
@@ -0,0 +1,680 @@
...
+        label_conf = process_conf['labels'] if 'labels' in process_conf else None
+        q = []
+        manager = multiprocessing.Manager()
+        return_dict = manager.dict()
+        for i, in_file in enumerate(in_files):
+            p = Process(target=parse_node_data, args=(i, in_file, feat_ops, node_id_col,
+                                                      label_conf, read_file, return_dict))
+            p.start()
+            q.append(p)
+            wait_process(q, num_processes)
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:486]: yes

File: tools/graph_loading/construct_graph.py (hunk 10 above)
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:662]: here i don't. we can add an argument to add reverse edges.

File: tools/graph_loading/construct_graph.py [hunk 18]
```diff
@@ -0,0 +1,680 @@
...
+    g = dgl.heterograph(edges, num_nodes_dict=num_nodes)
+    for ntype in node_data:
+        for name, data in node_data[ntype].items():
+            g.nodes[ntype].data[name] = th.tensor(data)
+    for etype in edge_data:
+        for name, data in edge_data[etype].items():
+            g.edges[etype].data[name] = th.tensor(data)
+
+    if args.output_format == "DistDGL":
+        dgl.distributed.partition_graph(g, args.graph_name, 1, args.output_dir, part_method="None")
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:671]: yes

File: tools/graph_loading/construct_graph.py [hunk 19]
```diff
@@ -0,0 +1,680 @@
...
+    argparser = argparse.ArgumentParser("Preprocess graphs")
+    argparser.add_argument("--conf_file", type=str, required=True,
+            help="The configuration file.")
+    argparser.add_argument("--num_processes", type=int, default=1,
+            help="The number of processes to process the data simulteneously.")
+    argparser.add_argument("--output_dir", type=str, required=True,
+            help="The path of the output data folder.")
+    argparser.add_argument("--graph_name", type=str, required=True,
+            help="The graph name")
+    argparser.add_argument("--remap_node_id", type=bool, default=False,
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:637]: the code checks the input node IDs. if they are already integers and are already sorted starting from 0, the script won't remap node IDs by default. if this argment is provided, the script will skip the checking and always remap node IDs. i wonder if we should provide users this argument.

File: tools/graph_loading/construct_graph.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:671]: i think we should provide users the option of partitions the input graph data.

File: tools/graph_loading/construct_graph.py [hunk 20]
```diff
@@ -0,0 +1,680 @@
...
+        feat_ops = parse_feat_ops(process_conf['features']) \
+                if 'features' in process_conf else None
+        label_conf = process_conf['labels'] if 'labels' in process_conf else None
+        q = []
+        manager = multiprocessing.Manager()
+        return_dict = manager.dict()
+        for i, in_file in enumerate(in_files):
+            p = Process(target=parse_edge_data, args=(i, in_file, feat_ops,
+                                                      src_id_col, dst_id_col, edge_type,
+                                                      node_id_map, label_conf,
```
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py:593]: here i assume all node types have node files. let's merge this PR first and add another PR to handle the case that a node type doesn't have node files. to handle this case, we need to add a complex implementation of computing unique node IDs. This PR is already very large.

[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py]: https://github.com/awslabs/graphstorm/issues/42

File: tools/graph_loading/construct_graph.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py]: support custom data split: https://github.com/awslabs/graphstorm/issues/41

File: tools/graph_loading/construct_graph.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py]: define rules for naming multipe output tensors: https://github.com/awslabs/graphstorm/issues/40

File: tests/end2end-tests/data_process/data_gen.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply tests/end2end-tests/data_process/data_gen.py:29]: Oh, it is test.

File: tools/graph_loading/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply tools/graph_loading/README.md]: Content.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 21]
```diff
@@ -0,0 +1,732 @@
...
+        dgl.save_graphs(os.path.join(args.output_dir, args.graph_name + ".dgl"), [g])
+    else:
+        raise ValueError('Unknown output format: {}'.format(args.output_format))
+    for ntype in node_id_map:
+        map_data = {}
+        map_data["orig"] = np.array(list(node_id_map[ntype].keys()))
+        map_data["new"] = np.array(list(node_id_map[ntype].values()))
+        write_data_parquet(map_data, os.path.join(args.output_dir, ntype + "_id_remap.parquet"))
+
+if __name__ == '__main__':
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:713]: Can library files have main function?

File: tools/graph_loading/construct_graph.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply tools/graph_loading/construct_graph.py]: yes

**Review by classicsong — APPROVED** [review]
